### DISCUSSION TITLE
Future of Spark

### DISCUSSION THREAD

**Discussion by westonwalker** [spark-dotnet/framework#24] [Ideas]
Hi all,

I wanted to share my thoughts on the future of spark and also get feedback/opinions from all of you.

### Blazor in .NET 8
Over the past week, I've played around with .NET 8 and the Blazor changes. My hope was I could use Blazor the same way Blazor Server works in .NET 7 but without the web socket dependency. Unfortunately, things like interactivity won't work unless you opt-in to Blazor Server (Websockets) or WASM.

This is a big deal because both of those have big problems. SignalR uses websockets. Servers can only handle so many websocket connections at a time.

WASM has a large download size and quite frankly, does not give a good user experience.

So I was left thinking, what should the future of Spark be knowing Blazor in .NET 8 isn't what I was hoping.

### Where to go from here
I'm thinking of make Spark a web project for .NET in general, not just Blazor.

This would mean 3 different project templates. All with the same backend and 3 choices for frontend.

The 3 frontend choices would be:
1. Blazor SSR (This is the new Blazor mode in .NET 8)
2. ASP.NET MVC
3. ASP.NET MVC with InertiaJS

If you haven't used InertiaJS, check it out here https://inertiajs.com/
The quick explanation is it's a small library to integrate React, Vue, or Svelte into your backend app while keeping the monolith architecture. All routing is done server side. So your MVC controllers would do some logic and return the frontend component that should render while maintaining all the benefits of using the frontend frameworks. It also means you can use all of the middleware provided by .NET, like authorization.

Providing the MVC option also gives devs who just want an API the ability to use all the Spark goodies like mailers, scheduled jobs, etc...

Let me know what you all think

**Comment by sturlath** [comment]
> Unfortunately, things like interactivity won't work unless you opt-in to Blazor Server (Websockets) or WASM.

Can you elaborate?

I take you are not as excited for the new Blazor as you were before?

**Reply by westonwalker** [reply]
I was hoping the SSR version of Blazor coming in .NET 8 worked just like Blazor Server works today but used AJAX calls instead of Websockets for it's interactivity.

Unfortunately, that is not the case.

This Counter component is a good example of interactivity. This would not work unless you opted-in for websockets or WASM.

![2023-07-07 21_26_01-Window](https://github.com/blazor-spark/blazor-spark/assets/66449990/1136e747-bcf6-4580-8317-8c7efa8503b3)

**Reply by sturlath** [reply]
So you are saying its to "heavy".. it either needs a steady SignalR connection (I think I have seen some tickets about much better reconnect story ) or a WASM download for something even this simple? So I guess if you have a hugely popular site you will either have to have a beefy SignalR hub or massive download (costly/slow).

Just trying to understand your reasoning for not liking it. Do you think that they could use AJAX or is that to big of a change?

**Reply by westonwalker** [reply]
Correct. You nailed my concerns for both WASM and Server. I know they could use AJAX, they just decided not too for interactivity. Libraries like Laravel Livewire are already doing this.

**Reply by sturlath** [reply]
Just had time to watch this https://youtu.be/iDLWv3xM1s0 and with the caviat that I havent done anything with this version, I think using rendermode == auto meets my needs. 

So for my frontend I would use serverside rendering for fast display and SEO friendly stuff. And then auto on the pages that have some dinamic on them. Downloading WASM in the background sounds fine if it then saves sockets later on.

Please correct me if I got this all wrong.

**Reply by westonwalker** [reply]
You are correct, that is how they are expecting us to use it in .NET 8. I haven't had a chance to test WASM in the .NET 8 preview. The user experience was pretty rough in my experience with WASM in .NET 7. Huge download sizes, app freezes for multiple seconds while it downloads. Hopefully that is fixed in .NET 8.

**Reply by jsalichos** [reply]
The key advantage of Blazor "United" is its ability to address the issues you mentioned by leveraging sockets during the WebAssembly (WASM) download process. This eliminates the need for an excessive number of SignalR circuits and prevents users from waiting idly while WASM is being downloaded.

Moreover, Blazor "United" offers additional benefits such as SSR mode and streaming rendering, which cater to SEO requirements or pages that do not require extensive interactivity.

With the introduction of .NET 8, Blazor reaching a level on par with other single-page application (SPA) frameworks..

**Reply by westonwalker** [reply]
That is the selling point.  I haven't seen any proof of those claims though and am very skeptical of how it will work in real world scenarios. There was also a talk done recently by Microsoft where they stated they would be splitting WASM project types back out to it's own thing. So it would seem like they are bailing on this idea.

**Reply by sturlath** [reply]
> So it would seem like they are bailing on this idea.

Do you have a link/info?

Maybe we can try to get some feedback on this from MS? Ticket og Twitter 😅

**Reply by jonathan-james-dev** [reply]
Have you not watched the latest video from Daniel Roth:
[https://www.youtube.com/watch?v=QD2-DwuOfKM](https://www.youtube.com/watch?v=QD2-DwuOfKM)

**Comment by SteveSimms** [comment]
Is there an inertia.js adapter for dotnet 6+? 
It would be cool if we could leverage SvelteKit with some of the goodness of Spark. Like auth and mailers.

**Reply by westonwalker** [reply]
Yes, in the coming weeks I'll be creating new templates for Spark. Right now the plan is to have an MVC template, MVC + Inertia template, and a Blazor template.

There is an open source Inertia adapter that has been abandoned if you'd like to check it out. It took me a while to get it working. https://github.com/Nothing-Works/inertia-aspnetcore

**Reply by ODumpling** [reply]
I tried the following adapter which is inspired off Nothing Works that was really easy to get working. https://github.com/kapi2289/InertiaCore

**Comment by smtm2033** [comment]
I would say finishing the first stable and complete version of Spark first, especially the back-end so that we can really start building something on it without concerns. Blazor server is okay for small scale application in anyway. Or you would move away from Blazor in the future?

**Reply by westonwalker** [reply]
Spark 1.0 will come out following the release of .NET 8 later this year.

I won't be moving away from Blazor. The plan is to support several project templates including Blazor SSR (coming in .NET 8), MVC, Razor Pages, and MVC API.

Blazor SSR can be changed to use Server or WASM very easily in .NET 8. So Spark will have a Blazor SSR template and the dev can change to WASM or Server if they wish.

**Reply by smtm2033** [reply]
thanks Weston. I can see Spark is drawing serious attention. If we look at the similar open-source projects and compare with the other technology stacks like PHP, JavaScript, Java, etc. in my opinion .NET core is far left behind. I believe Spark can take an important position especially when more and more contributors are joining.

**Comment by MarkCoffman** [comment]
I am sad to hear that they might be bailing on the idea.  If that is the
case I am not sure where blazor’s place is in the SPA space, and if not
there then in general.  I love blazor but the hot reload is a mess compared
to the various js frameworks and vite for example.  I love the syntax for
most concepts and the barrier to entry for concepts like state vs react is
lower.  But I must have missed that take about blazor united being bailed
on.

**Reply by westonwalker** [reply]
I've never been sold on Blazor being a true competitor to JS spa frameworks. To me it could shine as a server side templating language with much better component support than traditional razor views and pages.

The good news is most apps user interfaces aren't complex enough to even need React, Vue, or Angular. Server side rendering with something like HTMX can do most UI use cases without near the complexity a JS SPA framework will bring.

Also, if anyone is interested, the talk about separating Blazor project types is in the Spark discord.

**Comment by MarkCoffman** [comment]
Don’t sleep on blazor hybrid with Maui or WinForms.  It is legit.

**Reply by SteveSimms** [reply]
Blazor Maui bindings is nice. I wish they would focus on that more.

**Comment by westonwalker** [comment]
After taking in some feedback from the community and talking with some Microsoft devs about Blazor in .NET 8, I've made a final decision about the future of Spark.

Starting in .NET 8, Spark's main project type will be Blazor SSR. Any component within Blazor SSR can be tweaked to use Blazor Server, which gives you rich UI interactivity. This to me is a nice sweet spot.

Spark will continue to support the API, Razor and MVC templates. However, Blazor will see new Spark features first. I will also have to rely on community contributions more to help keep those other project templates up to date as it's hard to do all of them myself.

I'm sure some are wondering, what about WASM?

Well, the recent news is WASM will not fit right in with Blazor SSR and Blazor Server as previously proposed by Microsoft in .NET 8. It will be it's own separate project type. That is unfortunate but understandable, as implementing Auth and other things in WASM are completely different from Blazor SSR and Blazor Server. Because of this, Spark will not have first class support for WASM projects.

**Reply by ODumpling** [reply]
Any idea if you will implement a inertiajs template?

**Reply by westonwalker** [reply]
2 months ago I would have said yes. But after working quite a bit with Blazor in .NET 8 I'm leaning towards focusing on it. It is very good.

**Comment by MarkCoffman** [comment]
Would it be wise, then, to have your new template contain a Blazor Server
project and also a Microsoft.NET.Sdk.Razor client library project type by
default?  This project would contain all of the Pages, Components, and
Services. This way, if, down the road, someone wanted to add WASM, they
just reference the shared project?

You wouldn't need to have WASM out of the box, but should anyone add it,
they aren't moving a bunch of stuff into something that is accessible by
both.

I might still be thinking too much in .NET 6/7 terms, as I haven't explored
the .NET 8 features or proposed features, so maybe this shared UI (for lack
of a better term) project isn't necessary, but it feels like it would be.

**Reply by sturlath** [reply]
Any change you’r able to comment under the conversationa you are replying to? Its hard to follow 😅 just a suggestion

**Comment by pratikghag1089** [comment]
Oh, I love working with Blazor, particularly in web assembly. Your observation is accurate, though – the application size is exceptionally large, and it demands a robust machine for smooth operation. 

Have you considered the possibility of incorporating it with Angular or React?

**Reply by westonwalker** [reply]
I'm focusing mostly on Blazor SSR and Server at this point. That combination in .NET 8 can do anything you need in web development.

**Comment by hireflyapp** [comment]
Hi @westonwalker! Spark looks super interesting. Are there any plans for adding multi tenant support?
